// @jfschmiechen/react-event-calendar/components/event/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Paper = require('@material-ui/core/Paper');

var _Paper2 = _interopRequireDefault(_Paper);

require('./style.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Event = function Event(props) {
    var text = props.text,
        eventType = props.eventType,
        styleObj = props.styleObj,
        showArrowBefore = props.showArrowBefore,
        showArrowAfter = props.showArrowAfter,
        tileIndex = props.tileIndex,
        onClick = props.onClick,
        numTiles = props.numTiles;

    var heightOffset = numTiles > 35 ? 3.3 : 2.8; // Additional height of weekday divs
    var eventHeight = numTiles > 35 ? 0.9 : 1.2; // Divider
    return _react2.default.createElement(
        _Paper2.default,
        {
            className: eventType,
            onClick: onClick,
            style: {
                width: styleObj.width,
                height: tileIndex < 7 ? styleObj.height / eventHeight - heightOffset + '%' : styleObj.height / eventHeight + '%',
                backgroundColor: styleObj.color
            } },
        _react2.default.createElement(
            'span',
            {
                className: 'event-continues-arrow',
                style: {
                    display: showArrowBefore ? 'flex' : 'none',
                    float: 'left',
                    margin: 'auto -5px auto 10px'
                } },
            '<'
        ),
        _react2.default.createElement(
            'span',
            { className: 'event-text', style: { fontSize: '0.8em' } },
            '' + text
        ),
        _react2.default.createElement(
            'span',
            {
                className: 'event-continues-arrow',
                style: {
                    display: showArrowAfter ? 'flex' : 'none'
                } },
            '>'
        )
    );
};

exports.default = Event;
