// @jfschmiechen/react-event-calendar/components/calendar/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _calendarItem = require('../calendarItem');

var _calendarItem2 = _interopRequireDefault(_calendarItem);

require('./style.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Calendar = function Calendar(props) {
    var calendarList = props.calendarList,
        month = props.month,
        year = props.year,
        config = props.config;

    return _react2.default.createElement(
        'div',
        { className: 'calendar-container' },
        _react2.default.createElement(
            'div',
            { className: 'calendar-tile-container' },
            calendarList.map(function (eventArray, index) {
                return _react2.default.createElement(_calendarItem2.default, {
                    month: month,
                    year: year,
                    key: index,
                    eventArray: eventArray,
                    tileIndex: index,
                    numTiles: calendarList.length,
                    config: config
                });
            })
        )
    );
};

exports.default = Calendar;
